File: adt2unit/pas/release_sustaining_sound.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

procedure release_sustaining_sound(chan: Byte);
begin
  opl3out(_instr[02]+_chan_m[chan],63);
  opl3out(_instr[03]+_chan_c[chan],63);

  FillChar(fmpar_table[chan].adsrw_car,
           SizeOf(fmpar_table[chan].adsrw_car),0);
  FillChar(fmpar_table[chan].adsrw_mod,
           SizeOf(fmpar_table[chan].adsrw_mod),0);

  key_on(chan);
  opl3out(_instr[04]+_chan_m[chan],BYTE_NULL);
  opl3out(_instr[05]+_chan_c[chan],BYTE_NULL);
  opl3out(_instr[06]+_chan_m[chan],BYTE_NULL);
  opl3out(_instr[07]+_chan_c[chan],BYTE_NULL);

  key_off(chan);
  event_table[chan].instr_def := 0;
  reset_chan[chan] := TRUE;
end;

File: adt2unit/release_sustaining_sound.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

// chan: 1..20
void release_sustaining_sound (uint8_t chan) {
  chan--;

  opl3out (_instr[2] + _chan_m[chan], 63);
  opl3out (_instr[3] + _chan_c[chan], 63);

  memset (&fmpar_table[chan].adsrw_car, 0, sizeof (fmpar_table[0].adsrw_car));
  memset (&fmpar_table[chan].adsrw_mod, 0, sizeof (fmpar_table[0].adsrw_mod));

  key_on (chan + 1);
  opl3out (_instr[4] + _chan_m[chan], BYTE_NULL);
  opl3out (_instr[5] + _chan_c[chan], BYTE_NULL);
  opl3out (_instr[6] + _chan_m[chan], BYTE_NULL);
  opl3out (_instr[7] + _chan_c[chan], BYTE_NULL);

  key_off (chan + 1);
  event_table[chan].instr_def = 0;
  reset_chan[chan] = true;
}
